### PR TITLE
**Breaking:** Adjustments

### PR DESCRIPTION
**Pull request by TejasQ**
# Summary
This PR composes a bunch of fixes to various open issues (see commits). It contains [_one_ breaking change](https://github.com/contiamo/operational-ui/pull/788/files#diff-85f3197de18a071f349f7f5a257a93e3R25) to the `onClick` prop of `ContextMenu`: in terms of API, it only takes one argument: a `ContextMenuItem` instead of a `string | ContextMenuItem`.

# Related issue
#746 
#778 
#782 
#785 
#786 
#757 

# To be tested

Me
- [x] No error or warning in the console on `localhost:6060`

Tester 1

- [ ] Things look good on the demo.
  

Tester 2

- [ ] Things look good on the demo.

### CODE REVIEW

File: src/Autocomplete/util/makeItems.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { IContextMenuItem } from "../../ContextMenu/ContextMenu.Item"
```
Comment (peterszerzo) [inline src/Autocomplete/util/makeItems.ts]: Does this method need to be in its own folder in a separate file? Other components use the `ComponentName/ComponentName.utils.ts` pattern, where utility methods are piled up (see e.g. `DatePicker`).

File: src/ContextMenu/ContextMenu.tsx [hunk 2]
```diff
@@ -37,6 +37,7 @@ export interface ContextMenuProps extends DefaultProps {
 export interface State {
   isOpen: boolean
   focusedItemIndex: number
+  items: IContextMenuItem[]
```
Comment (peterszerzo) [inline src/ContextMenu/ContextMenu.tsx]: Is there another reason this is in state other than a way to convert between `string`'s and `Item`'s? If not, converting from `string[]` to `Item[]` should happen on the fly in render in order to avoid duplicating state.

File: src/ContextMenu/ContextMenu.tsx [hunk 3]
```diff
@@ -103,16 +104,34 @@ class ContextMenu extends React.Component<ContextMenuProps, State> {
...
     focusedItemIndex: 0,
+    items: [],
   }
 
   public static defaultProps: Partial<ContextMenuProps> = {
     align: "left",
     embedChildrenInMenu: false,
   }
 
+  /**
```
Comment (peterszerzo) [inline src/ContextMenu/ContextMenu.tsx]: See comment above in this file.

File: src/HeaderMenu/HeaderMenu.tsx [hunk 4]
```diff
@@ -5,7 +5,7 @@ import styled from "../utils/styled"
 
 export interface HeaderMenuProps extends DefaultProps {
   /** Clickable component(s) from which menu appears  */
-  children: React.ReactNode[]
+  children: React.ReactNode | React.ReactNode[]
```
Comment (peterszerzo) [inline src/HeaderMenu/HeaderMenu.tsx]: What is the reason behind this change?

File: src/ContextMenu/ContextMenu.tsx [hunk 5]
```diff
@@ -103,16 +104,34 @@ class ContextMenu extends React.Component<ContextMenuProps, State> {
...
+   */
+
+  public static getDerivedStateFromProps(props: ContextMenuProps) {
+    return {
+      items: props.items.map(item => {
+        if (typeof item === "string") {
+          return { label: item }
+        } else {
+          return item
+        }
```
Comment (ImogenF) [inline src/ContextMenu/ContextMenu.tsx]: This would be nicer as a ternary operation.

File: src/Autocomplete/util/makeItems.ts (hunk 1 above)
Comment (TejasQ) [reply src/Autocomplete/util/makeItems.ts]: Sure! I can do that!

File: src/ContextMenu/ContextMenu.tsx (hunk 2 above)
Comment (TejasQ) [reply src/ContextMenu/ContextMenu.tsx]: My reasoning here is because `state.items` is actually _derived state_ from the props given to this component. Consider,

`<ContextMenu items={["your", "mom", "is", "amazing"]}`

those are initial props. We then [derive the state](https://github.com/contiamo/operational-ui/blob/f450ad06dc3b17016912ee00c801ef27bfd64071/src/ContextMenu/ContextMenu.tsx#L123), which _cannot and should not_ be plain strings, but objects that implement the `IContextMenuItem` interface.

That's the chief reason I put them in state: it communicates intent (to me) a little more clearly. And from that point on, `render` just renders the state: it is essentially a function of the `state` without side effects, maintaining its purity.

File: src/ContextMenu/ContextMenu.tsx (hunk 3 above)
Comment (TejasQ) [reply src/ContextMenu/ContextMenu.tsx]: [Responded.](https://github.com/contiamo/operational-ui/pull/788#discussion_r225105596)

File: src/HeaderMenu/HeaderMenu.tsx (hunk 4 above)
Comment (TejasQ) [reply src/HeaderMenu/HeaderMenu.tsx]: Previously, one could not have a single child in `HeaderMenu`. This became problematic when you wanted `<Icon name="MyIcon" /> My Project` as a child. There is [more info in the related issue](https://github.com/contiamo/operational-ui/issues/778).

File: src/ContextMenu/ContextMenu.tsx (hunk 2 above)
Comment (peterszerzo) [reply src/ContextMenu/ContextMenu.tsx]: This isn't derived state, it is simply props transformed into something else one-to-one. As in the docs https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops, the `getDerivedStateFromProps` method is intended for pieces of state that depend on changes in props over time e.g. setting intermediate states in transitions.

I also disagree that the intent is clearer: putting things in local state sends a false signal to another developer that the state contains additional information than what is passed in from props, or worse, that it is alterable by `this.setState`, whereas it should at all times just reflect what it has been given through props.

Comment (TejasQ) [reply src/ContextMenu/ContextMenu.tsx]: You make a lot of sense. Thanks for pointing this out. Will refactor!

File: src/ContextMenu/ContextMenu.tsx [hunk 6]
```diff
@@ -103,7 +103,10 @@ class ContextMenu extends React.Component<ContextMenuProps, State> {
     }
   }
 
-  public state: State = {
+  private makeItem = (itemFromProps: ContextMenuProps["items"][-1]) =>
```
Comment (peterszerzo) [inline src/ContextMenu/ContextMenu.tsx]: Can we add the comment from before, the one that said that the `getDerivedStateFromProps` solution was meant to standardize the public API, only that here we're achieving the same thing differently?
